Reduce the number of custom set implementations

zmb3 opened the pull request.

zmb3 committed:
Reduce the number of custom set implementations

zmb3 committed:
Move set to a dedicated package

lib/utils is a large package and we don't want to force all packages
that need a set to import everything else in there.

A small alias remains in lib/utils to avoid breakign teleport.e.

creack commented on lib/utils/set/set_test.go:
```diff
@@ -0,0 +1,316 @@
+		t.Run(test.name, func(t *testing.T) {
+			s := New(test.elements...)
+
+			var iteratedElements []string
```
nit

```suggestion
			iteratedElements := make([]string, 0, len(s))
```

creack commented on lib/utils/set/set.go:
```diff
@@ -0,0 +1,128 @@
+//
+// Prefer using the natural range iteration where possible
+func (s Set[T]) Elements() []T {
+	elements := make([]T, 0, len(s))
```
May be better for a different PR, but could modernize.

```suggestion
	return slices.Collect(maps.Keys(elements))
```

zmb3 committed:
Fix tests

Some tests were depending on sorted elements, others cared about
the difference between a nil slice and a non-nil but empty slice.
